src/V3Randomize.cpp

```diff
@@ -1303,6 +1303,71 @@ class RandomizeVisitor final : public VNVisitor {
         UINFO(9, "created " << varp << endl);
         return newp;
     }
+    AstNodeStmt* createArrayForeachLoop(FileLine* const fl, AstNodeDType* const dtypep,
+                                        AstNodeExpr* exprp) {
+        V3UniqueNames* uniqueNamep = new V3UniqueNames{"__Vrandarr"};
+        AstNodeDType* tempDTypep = dtypep;
+        AstVar* randLoopIndxp = nullptr;
+        AstNodeStmt* stmtsp = nullptr;
+
+        if (VN_CAST(tempDTypep, DynArrayDType)) {
+            AstCMethodHard* tempElementp = nullptr;
+            while (VN_CAST(tempDTypep, DynArrayDType)) {
+                AstVar* const newRandLoopIndxp
+                    = new AstVar{fl, VVarType::VAR, uniqueNamep->get(""),
+                                 dtypep->findBasicDType(VBasicDTypeKwd::UINT32)};
+                randLoopIndxp = AstNode::addNext(randLoopIndxp, newRandLoopIndxp);
+                tempElementp
+                    = tempElementp
+                          ? new AstCMethodHard{fl, tempElementp, "atWrite",
+                                               new AstVarRef{fl, newRandLoopIndxp, VAccess::READ}}
+                          : new AstCMethodHard{fl, exprp, "atWrite",
+                                               new AstVarRef{fl, newRandLoopIndxp, VAccess::READ}};
+                tempElementp->dtypep(tempDTypep->subDTypep());
+                tempDTypep = tempDTypep->virtRefDTypep();
+            }
+            if (VN_IS(tempElementp->dtypep()->skipRefp(), StructDType)) {
+                tempElementp->v3warn(
+                    E_UNSUPPORTED,
+                    "Unsupported: CreateArrayForeachLoop currently does not support "
+                    "this data type. (Struct-Array unconstrained "
+                    "randomization is not fully supported)");
+            }
+            AstSelLoopVars* const randLoopVarp
+                = new AstSelLoopVars{fl, exprp->cloneTree(false), randLoopIndxp};
+            stmtsp = new AstForeach{fl, randLoopVarp, newRandStmtsp(fl, tempElementp, nullptr)};
+        } else if (VN_CAST(tempDTypep, UnpackArrayDType)) {
+            AstArraySel* tempElementp = nullptr;
+            while (VN_CAST(tempDTypep, UnpackArrayDType)) {
+                AstVar* const newRandLoopIndxp
+                    = new AstVar{fl, VVarType::VAR, uniqueNamep->get(""),
+                                 dtypep->findBasicDType((VBasicDTypeKwd::UINT32))};
+                randLoopIndxp = AstNode::addNext(randLoopIndxp, newRandLoopIndxp);
+                tempElementp
+                    = tempElementp
+                          ? new AstArraySel{fl, tempElementp,
+                                            new AstVarRef{fl, newRandLoopIndxp, VAccess::READ}}
+                          : new AstArraySel{fl, exprp,
+                                            new AstVarRef{fl, newRandLoopIndxp, VAccess::READ}};
```

Comment:
Note only the subexpression differs between branches. Same above.
```suggestion
                    = new AstArraySel{fl, tempElementp ? tempElementp : exprp,
                                            new AstVarRef{fl, newRandLoopIndxp, VAccess::READ}};
```